shell/packages/sandstorm-db/db.js

```diff
@@ -560,6 +566,52 @@ isSignedUpOrDemo = function () {
   return false;
 }
 
+var calculateReferralBonus = function(accountId, plan) {
+  // This function returns an object of the form:
+  //
+  // - {grains: 0, storage: 0}
+  //
+  // which are extra resources this account gets as part of participating in the referral
+  // program. (Storage is measured in bytes, as usual for plans.)
+
+
+  // Authorization note: Only call this if accountId is the current user!
+  var isPaid = (plan && plan !== "free");
+
+  successfulReferralsCount = countReferrals(accountId);
+  if (isPaid) {
+    var maxPaidStorageBonus = 30 * 1e9;
+    return {grains: 0,
+            storage: Math.min(
+              successfulReferralsCount * 2 * 1e9,
+              maxPaidStorageBonus)};
+  } else {
+    var maxFreeStorageBonus = 2 * 1e9;
+    return {grains: successfulReferralsCount * Infinity,
+            storage: Math.min(
+              successfulReferralsCount * 50 * 1e6,
+              maxFreeStorageBonus)};
+  }
+}
+
+var countReferrals = function (accountId) {
+  var referredIdentityIds = Meteor.users.findOne(
+    {_id: accountId},
+    {fields: {referredIdentityIds: 1}}).referredIdentityIds;
+  return (referredIdentityIds && referredIdentityIds.length || 0);
+}
+
+getUserQuota = function (user) {
+  // Re-fetch the user, since typically Meteor gives is an object with just an _id.
```

Comment:
Eh? I've never known that to be true.

`Meteor.user()` on the client side will only return published fields, but on the server side it should return the whole document.